galsim/fft.py

```diff
@@ -0,0 +1,267 @@
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+"""@file fft.py
+Functional equivalents of (some of) the np.fft functions, but using FFTW.
+
+These should be drop in replacements for np.fft.* functions.  e.g.
+
+    >>> karray = galsim.fft.fft2(xarray)
+
+is functionally equivalent to
+
+    >>> karray = np.fft.fft2(xarray)
+
+but should be a bit faster.
+
+Note that the GalSim versions often only implement the normal use case without many of the
+advanced options available with the numpy functions.  This is mostly laziness on our part --
+we only implemented the functions that we needed.  If your usage requires some option available
+in the numpy version, feel free to post a feature request on our GitHub page.
+"""
+
+import galsim
+import numpy as np
+
+def fft2(a, shift_in=False, shift_out=False):
+    """Compute the 2-dimensional discrete Fourier Transform.
+
+    For valid inputs, the result is equivalent to numpy.fft.fft2(a), but usually faster.
+
+        >>> ka1 = numpy.fft.fft2(a)
+        >>> ka2 = galsim.fft.fft2(a)
+
+    Restrictions on this version vs the numpy version:
+
+        - The input array must be 2-dimensional.
+        - The size in each direction must be even. (Ideally 2^k or 3*2^k for speed, but this is
+          not required.)
+        - If it has a real dtype, it will be coerced to numpy.float64.
+        - If it hsa a complex dtype, it will be coerced to numpy.complex128.
+
+    The returned array will be complex with dtype numpy.complex128.
+
+    If shift_in is True, then this is equivalent to applying numpy.fft.fftshift to the input.
+
+        >>> ka1 = numpy.fft.fft2(numpy.fft.fftshift(a))
+        >>> ka2 = galsim.fft.fft2(a, shift_in=True)
+
+    If shift_out is True, then this is equivalent to applying numpy.fft.fftshift to the output.
+
+        >>> ka1 = numpy.fft.fftshift(numpy.fft.fft2(a))
+        >>> ka2 = galsim.fft.fft2(a, shift_out=True)
+
+    @param a            The input array to be transformed
+    @param shift_in     Whether to shift the input array so that the center is moved to (0,0).
+    @param shift_out    Whether to shift the output array so that the center is moved to (0,0).
+
+    @returns a complex numpy array
+    """
+    s = a.shape
+    if len(s) != 2:
+        raise ValueError("Input array must be 2D.")
+    M, N = s
+    Mo2 = M // 2
+    No2 = N // 2
+
+    if M != Mo2*2 or N != No2*2:
+        raise ValueError("Input array must have even sizes")
+
+    if a.dtype.kind == 'c':
+        a = a.astype(np.complex128, copy=False)
+        xim = galsim._galsim.ConstImageViewC(a, -No2, -Mo2)
+        kar = xim.cfft(shift_in=shift_in, shift_out=shift_out).array
+    else:
+        a = a.astype(np.float64, copy=False)
+        xim = galsim._galsim.ConstImageViewD(a, -No2, -Mo2)
+        if False:
```

Comment:
No, I'll remove it.  It was useful for testing to flip between them to see the speed difference.  The rfft2 version is a good 30% or so faster IIRC.